staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go

```diff
@@ -0,0 +1,106 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"k8s.io/klog/v2"
+	"net/http"
+	"time"
+
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+var (
+	// The 'timeout' query parameter in the request URL has an invalid timeout specifier
+	errInvalidTimeoutInURL = errors.New("invalid timeout specified in the request URL")
+
+	// The timeout specified in the request URL exceeds the global maximum timeout allowed by the aiserver.
+	errTimeoutExceedsMaximumAllowed = errors.New("timeout specified in the request URL exceeds the maximum timeout allowed by the apiserver")
+)
+
+// WithRequestDeadline determines the deadline of the given request and sets a new context
+// with the appropriate timeout.
+// timeout specifies the default request timeout value
+func WithRequestDeadline(handler http.Handler, longRunning request.LongRunningRequestCheck, requestTimeoutMaximum time.Duration) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+
+		requestInfo, ok := request.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, req, http.StatusInternalServerError, fmt.Errorf("no RequestInfo found in context, handler chain must be wrong"))
+			return
+		}
+		if longRunning(req, requestInfo) {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		userSpecifiedTimeout, ok, err := parseTimeout(req, requestTimeoutMaximum)
+		if err != nil {
+			switch {
+			case err == errInvalidTimeoutInURL || err == errTimeoutExceedsMaximumAllowed:
+				handleError(w, req, http.StatusBadRequest, err)
+			default:
+				handleError(w, req, http.StatusInternalServerError, err)
+			}
+
+			return
+		}
+
+		timeout := requestTimeoutMaximum
+		if ok {
+			timeout = userSpecifiedTimeout
+		}
+
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
+
+		req = req.WithContext(ctx)
+		handler.ServeHTTP(w, req)
+	})
+}
+
+// parseTimeout parses the given HTTP request URL and extracts the timeout query parameter
+// value if specified by the user.
+// If a timeout is not specified or the value specified is malformed then
+// the function returns false.
+func parseTimeout(req *http.Request, maximumTimeoutAllowed time.Duration) (time.Duration, bool, error) {
+	value := req.URL.Query().Get("timeout")
+	if value == "" {
+		return 0, false, nil
+	}
+
+	timeout, err := time.ParseDuration(value)
+	if err != nil {
+		return 0, false, errInvalidTimeoutInURL
+	}
+
+	if timeout > maximumTimeoutAllowed {
```

Comment:
I'd prefer this check in the function that calls parseTimeout, so that parsing only parses.

Reply:
done